### DISCUSSION TITLE
Usage of gatsby-ts in Gatsby Starter Dogmatism

### DISCUSSION THREAD

**Discussion by openscript** [Js-Brecht/gatsby-plugin-ts-config#43] [Show and tell]
Hello folks

Recently I started quite a few projects with Gatsby. Some of them are already using the v2. So I thought I'll create a starter, which reflect my usual set up. Maybe it helps others. 

https://github.com/openscript/gatsby-starter-dogmatism

Thank you @Js-Brecht for all the work! It helps so much to keep `gatsby-node` and `gatsby-config` nice, clean and of course typed.

**Comment by Js-Brecht** [comment]
@openscript I just released the new version called `gatsby-ts` as a release candidate.  It gives you a CLI to call gatsby with, which allows you to define `gatsby-config.ts` and `gatsby-node.ts` in the root of your repository.

Check out the docs [here](https://github.com/Js-Brecht/gatsby-plugin-ts-config/tree/gatsby-ts#readme).

**Reply by openscript** [reply]
Great! With that it gets cleaner. I've updated the starter.

Unfortunately many of the plugins don't provide typings for their `PluginOptions`. Is there a way to generate it from their Joi schemes?

**Reply by Js-Brecht** [reply]
Yeah, it is unfortunate that a lot of plugin authors don’t write types for their plugin’s configuration.  I imagine it’s mostly due to the fact that the vast majority of them don’t even write their plugin in Typescript.

There might be a generator out there that could generate types from a Joi schema.  I know there’s TS transformers that generate a schema from types, so who knows.  Could be worth looking into

**Reply by openscript** [reply]
Maybe [this](https://github.com/mrjono1/joi-to-typescript) is also an option.

**Reply by Js-Brecht** [reply]
That's pretty neat.  Maybe the plugin authors will accept some pull requests to add types to their projects

**Reply by openscript** [reply]
Or we do something like https://github.com/DefinitelyTyped/DefinitelyTyped

**Comment by openscript** [comment]
I'm pretty sad about this: https://github.com/gatsbyjs/gatsby/discussions/33736
Now I've to write types and think of something so a plugin can provide it too..

**Reply by Js-Brecht** [reply]
It looks like they're talking about the GraphQL types.  The Typescript types are still good, AFAICT:

* For pages: [PageProps](https://github.com/gatsbyjs/gatsby/blob/eb3db0591ccac51194b890d88945a9c6619a5466/packages/gatsby/index.d.ts#L80-L84)
* For `gatsby-browser`/`gatsby-ssr`: [WrapPageElementBrowserArgs](https://github.com/gatsbyjs/gatsby/blob/eb3db0591ccac51194b890d88945a9c6619a5466/packages/gatsby/index.d.ts#L1473-L1480)
* For accessing the `wrapPageElement` API type in `gatsby-browser` and `gatsby-ssr`:
  * [GatsbyBrowser](https://github.com/gatsbyjs/gatsby/blob/eb3db0591ccac51194b890d88945a9c6619a5466/packages/gatsby/index.d.ts#L542-L546)
  * [GatsbySSR](https://github.com/gatsbyjs/gatsby/blob/eb3db0591ccac51194b890d88945a9c6619a5466/packages/gatsby/index.d.ts#L608-L611)
* For use in `createPages` API in `gatsby-node: [createPage](https://github.com/gatsbyjs/gatsby/blob/eb3db0591ccac51194b890d88945a9c6619a5466/packages/gatsby/index.d.ts#L1108-L1113)

Unless I'm misunderstanding what types you're using

**Reply by openscript** [reply]
Sorry, that I didn't elaborate on this. I'm only talking about the `SitePageContext`. In Gatsby V4 the `PageContext` is just a `JSON` field and the types are not inferenced anymore. As they state in their [migration guide](https://v4.gatsbyjs.com/docs/reference/release-notes/migrating-from-v3-to-v4/#changes-to-built-in-types):

> The built-in type SitePage now returns the pageContext key as JSON and won’t infer any other information anymore. The SitePlugin type now has two new keys: pluginOptions: JSON and packageJson: JSON.

So no more types for the `PageContext` can be generated from the GraphQL schema.

**Reply by Js-Brecht** [reply]
Oh right... yeah, it is odd they would do that.
